File: src/main/java/com/github/dkoval/leetcode/challenge/RemoveNthNodeFromEndOfList.java
package com.github.dkoval.leetcode.challenge;

import com.github.dkoval.leetcode.ListNode;

/**
 * <a href="https://leetcode.com/problems/remove-nth-node-from-end-of-list/">Remove Nth Node From End of List</a>
 * <p>
 * Given the head of a linked list, remove the nth node from the end of the list and return its head.
 * <p>
 * Follow up: Could you do this in one pass?
 * <p>
 * Constraints:
 * <ul>
 *  <li>The number of nodes in the list is sz.</li>
 *  <li>1 <= sz <= 30.</li>
 *  <li>0 <= Node.val <= 100.</li>
 *  <li>1 <= n <= sz.</li>
 * </ul>
 */
public interface RemoveNthNodeFromEndOfList {

    ListNode removeNthFromEnd(ListNode head, int n);

    // O(N) time | O(1) space
    class RemoveNthNodeFromEndOfListRev1 implements RemoveNthNodeFromEndOfList {

        @Override
        public ListNode removeNthFromEnd(ListNode head, int n) {
            // advance `fast` pointer by n steps forward
            ListNode fast = head;
            while (n-- > 0) {
                fast = fast.next;
            }

            // now, keep on shifting `slow` and `fast `pointers until `fast` reaches the end of the list;
            // at the time `fast` reaches the end of the list, `slow` will point at the n-th node from the end of the list
            ListNode slow = head;
            ListNode prev = null;
            while (fast != null) {
                prev = slow;
                slow = slow.next;
                fast = fast.next;
            }

            // finally, remove n-th node from the end of the list
            if (prev != null) {
                prev.next = slow.next;
                return head;
            }
            // corner case: 1st node of the list is to be removed
            return head.next;
        }
    }
}

File: src/main/kotlin/com/github/dkoval/leetcode/interview/linkedlist/RemoveNthNodeFromEndOfList.kt
package com.github.dkoval.leetcode.interview.linkedlist

import com.github.dkoval.leetcode.ListNode

/**
 * [Remove Nth Node From End of List](https://leetcode.com/explore/featured/card/top-interview-questions-easy/93/linked-list/603/)
 *
 * Given a linked list, remove the n-th node from the end of list and return its head.
 */
object RemoveNthNodeFromEndOfList {

    fun removeNthFromEnd(head: ListNode?, n: Int): ListNode? {
        // Add an auxiliary "dummy" node, which points to the list head.
        // The "dummy" node is used to simplify some corner cases such as a list with only one node,
        // or removing the head of the list.
        val dummy = ListNode(42)
        dummy.next = head
        var slow: ListNode? = dummy
        var curr: ListNode? = dummy

        // advance `curr` by n + 1 steps to create n nodes gap between `curr` and `slow` pointers
        repeat(n + 1) {
            curr = curr?.next
        }

        // move `curr` to the end, maintaining the gap
        while (curr != null) {
            curr = curr?.next
            slow = slow?.next
        }
        slow?.next = slow?.next?.next
        return dummy.next
    }
}